### PR TITLE
Fix S2259 FP: SE engine doesn't take into account element existence collection methods

### LINKED ISSUE #8266: Fix S2259 FP: SE engine doesn't take into account element existence collection methods
From this [community post](https://community.sonarsource.com/t/upgrade-sonarqube-9-9-to-10-2-now-see-more-bugs-when-reviewing-the-c-code/103255/3).

This might be fixed as part of #7457:
> Learn from well-known (extension-) methods like list.Any()

```cs
void Method(Exception[] array)
{
    if (array.Any())
    {
        Exception exception = array.FirstOrDefault();
        Console.WriteLine(exception.Message); // Noncompliant - FP if Any is true, FirstOrDefault will never return null
    }
}
```

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Fixes #8266

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.Enumerable.cs [hunk 1]
```diff
@@ -125,9 +136,20 @@ private static ProgramState[] ProcessElementExistsCheckMethods(ProgramState stat
...
-            (symbol.IsExtensionMethod && symbol.Parameters.Length == 1)
-            || symbol.Parameters.IsEmpty;
     }
+
+    private static bool HasParameters(IMethodSymbol symbol) =>
+        symbol.IsExtensionMethod switch
+        {
+            true => symbol.Parameters.Length > 1,
+            false => !symbol.Parameters.IsEmpty
+        };
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.Enumerable.cs:146]:
```suggestion
        symbol.IsExtensionMethod
          ? symbol.Parameters.Length > 1
          :  !symbol.Parameters.IsEmpty;
```

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM (review was mostly done sync with @mary-georgiou-sonarsource and @Tim-Pohlmann)

File: analyzers/tests/SonarAnalyzer.Test/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 2]
```diff
@@ -1028,6 +1028,21 @@ public void Invocation_ElementOrDefault_CollectionNotEmpty_ReferenceElementType(
...
+    [DataRow("SingleOrDefault()")]
+    public void Invocation_ElementOrDefault_CollectionEmpty_ReferenceElementType(string expression)
+    {
+        var code = $"""
+            collection.Clear();
+            var value = collection.{expression};
+            Tag("Value", value);
+            """;
+        var validator = SETestContext.CreateCS(code, $"List<object> collection").Validator;
+        validator.TagValue("Value").Should().HaveOnlyConstraint(ObjectConstraint.Null);
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:1043]: You can merge these tests into one if you extract the collection type into a parameter:
- `"List<object>"`
- `"Dictionary<int, object>"`

Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:1043]: yes but I'm testing different things - in one case it's a reference that can return null and in the other case it never returns cause the keyvalue pair is a struct.

**Pull request merged**